File: MATLAB/UnicornConnect.m
function [s] = UnicornConnect(device, timeout)
% This function connects to a Unicorn device through a specified port and initiates data acquisition.
% Input: 
% - 'device' (string): which is the name of the serial port connected to the device (e.g., 'COM7')
% - 'timeout' (number): which is the time in seconds for the connection timeout
% Output:
% s (serialport object) - a serial port object representing the open connection to the Unicorn device

% UNICORN Bluetooth commands
start_acq =         [0x61, 0x7C, 0x87]';
start_response =    [0x00, 0x00, 0x00]';


 s = serialport(device, 115200, 'Timeout', timeout);
        disp(['Connected to serial port ', device]);
        fopen(s);
        if strcmp(s.Status, 'open')
            fwrite(s, start_acq);
        else
            error('Port is not open');
        end

        disp(['Started Unicorn on port ', device]);

        % check if first 3 bytes are 0
        payload = fread(s, 3, 'uint8');
        if ~isequal(payload, start_response)
            error(append('Unsuccessful start data stream on port:', device));
        end

File: Python/UNICORNConnect.py
import serial

class UNICORNDevice:
    def __init__(self, device, timeout):
        self.device = device
        self.timeout = timeout
        self.start_acq = [0x61, 0x7C, 0x87]
        self.start_response = [0x00, 0x00, 0x00]
        self._ser = None  # Use _ prefix to indicate it should not be accessed directly

    def connect(self):
        # Method to connect to the device
        self._ser = serial.Serial(self.device, 115200, timeout=self.timeout)
        self._ser.reset_input_buffer()
        self._ser.write(bytearray(self.start_acq))

        payload = self._ser.read(3)

        if payload == bytearray(self.start_response):
            print('Connection with Unicorn successful!')
        else:
            raise Exception(f'Unsuccessful start data stream on port: {self.device}')
            

    def serial_conn(self):
        # Method to access the serial port object
        if self._ser is None:
            raise Exception('Device not yet connected. Use connect() method first.')
        return self._ser
